File: web/src/slash-commands.ts
import type { Buffer } from "./app-state";

export type SlashCommand = {
  cmd: string;
  args: string;
  desc: string;
};

type SendCmd = (cmd: Record<string, unknown>) => void;

type SlashCommandHandler = (ctx: { buffer: Buffer; rest: string[]; args: string; sendCmd: SendCmd }) => void;

type SlashCommandEntry = SlashCommand & {
  names: string[];
  aliases?: SlashCommand[];
  handle: SlashCommandHandler;
};

function command(cmd: string, args: string, desc: string, handle: SlashCommandHandler): SlashCommandEntry {
  return { cmd, args, desc, names: [cmd.slice(1)], handle };
}

function alias(
  cmd: string,
  aliasName: string,
  args: string,
  desc: string,
  handle: SlashCommandHandler,
  aliasDesc = `${desc} (alias)`,
): SlashCommandEntry {
  return {
    cmd,
    args,
    desc,
    names: [cmd.slice(1), aliasName],
    aliases: [{ cmd: `/${aliasName}`, args, desc: aliasDesc }],
    handle,
  };
}

function networkCommand(type: string, cmd: string, args: string, desc: string): SlashCommandEntry {
  return command(cmd, args, desc, ({ buffer, args, sendCmd }) => {
    sendCmd({ type, network_id: buffer.network_id, content: args });
  });
}

function networkTargetCommand(type: string, cmd: string, args: string, desc: string): SlashCommandEntry {
  return command(cmd, args, desc, ({ buffer, args, sendCmd }) => {
    sendCmd({ type, network_id: buffer.network_id, target: args });
  });
}

function bufferCommand(type: string, cmd: string, args: string, desc: string): SlashCommandEntry {
  return command(cmd, args, desc, ({ buffer, args, sendCmd }) => {
    sendCmd({ type, buffer_id: buffer.id, content: args });
  });
}

function bufferTargetCommand(type: string, cmd: string, args: string, desc: string): SlashCommandEntry {
  return command(cmd, args, desc, ({ buffer, args, sendCmd }) => {
    sendCmd({ type, buffer_id: buffer.id, target: args });
  });
}

const registry: SlashCommandEntry[] = [
  command("/join", "<channel>", "Join a channel", ({ buffer, args, sendCmd }) => {
    sendCmd({ type: "join", network_id: buffer.network_id, channel: args });
  }),
  bufferCommand("part", "/part", "[reason]", "Leave current channel"),
  command("/msg", "<nick> <text>", "Send private message", ({ buffer, rest, sendCmd }) => {
    const [target, ...msgRest] = rest;
    if (target) {
      sendCmd({ type: "msg", network_id: buffer.network_id, target, content: msgRest.join(" ") });
    }
  }),
  bufferCommand("me", "/me", "<action>", "Send /me action"),
  networkCommand("nick", "/nick", "<newnick>", "Change your nick"),
  bufferCommand("topic", "/topic", "[new topic]", "View or set channel topic"),
  networkTargetCommand("whois", "/whois", "<nick>", "Query user info"),
  command("/invite", "<nick> [channel]", "Invite nick to channel", ({ buffer, rest, sendCmd }) => {
    const [target, channel] = rest;
    sendCmd({ type: "invite", network_id: buffer.network_id, target, channel: channel || buffer.name });
  }),
  command("/kick", "<nick> [reason]", "Kick nick from channel", ({ buffer, rest, sendCmd }) => {
    const [target, ...reason] = rest;
    sendCmd({ type: "kick", buffer_id: buffer.id, target, content: reason.join(" ") });
  }),
  bufferCommand("mode", "/mode", "<modes> [params]", "Set channel modes"),
  networkCommand("raw", "/raw", "<line>", "Send raw IRC line"),
  networkCommand("away", "/away", "[message]", "Set away status"),
  command("/back", "", "Clear away status", ({ buffer, sendCmd }) => {
    sendCmd({ type: "back", network_id: buffer.network_id });
  }),
  networkCommand("quit", "/quit", "[message]", "Disconnect from network"),
  alias(
    "/rejoin",
    "cycle",
    "",
    "Rejoin current channel",
    ({ buffer, sendCmd }) => {
      sendCmd({ type: "rejoin", buffer_id: buffer.id });
    },
    "Rejoin current channel (alias)",
  ),
  command("/notice", "<target> <text>", "Send a NOTICE", ({ buffer, rest, sendCmd }) => {
    const [target, ...msgRest] = rest;
    if (target) {
      sendCmd({ type: "notice", network_id: buffer.network_id, target, content: msgRest.join(" ") });
    }
  }),
  command("/ctcp", "<nick> <cmd> [args]", "Send CTCP request", ({ buffer, rest, sendCmd }) => {
    const [target, ctcpCmd, ...ctcpArgs] = rest;
    if (target && ctcpCmd) {
      sendCmd({
        type: "ctcp",
        network_id: buffer.network_id,
        target,
        content: ctcpCmd + (ctcpArgs.length > 0 ? ` ${ctcpArgs.join(" ")}` : ""),
      });
    }
  }),
  networkTargetCommand("query", "/query", "<nick>", "Open PM buffer without messaging"),
  networkCommand("list", "/list", "[filter]", "List channels on server"),
  bufferTargetCommand("op", "/op", "<nick>", "Give channel op (+o)"),
  bufferTargetCommand("deop", "/deop", "<nick>", "Remove channel op (-o)"),
  bufferTargetCommand("voice", "/voice", "<nick>", "Give voice (+v)"),
  bufferTargetCommand("devoice", "/devoice", "<nick>", "Remove voice (-v)"),
  bufferTargetCommand("ban", "/ban", "<mask>", "Ban mask (+b)"),
  bufferTargetCommand("unban", "/unban", "<mask>", "Unban mask (-b)"),
  command("/banlist", "", "Show channel ban list", ({ buffer, sendCmd }) => {
    sendCmd({ type: "banlist", buffer_id: buffer.id });
  }),
  command("/kickban", "<nick> [reason]", "Kick and ban nick", ({ buffer, rest, sendCmd }) => {
    const [target, ...reason] = rest;
    if (target) {
      sendCmd({ type: "kickban", buffer_id: buffer.id, target, content: reason.join(" ") });
    }
  }),
  networkTargetCommand("ignore", "/ignore", "<nick>", "Ignore nick or mask"),
  networkTargetCommand("unignore", "/unignore", "<nick>", "Remove ignore"),
  command("/ignorelist", "", "List ignored masks", ({ buffer, sendCmd }) => {
    sendCmd({ type: "ignorelist", network_id: buffer.network_id });
  }),
];

const registryByName = new Map(registry.flatMap((entry) => entry.names.map((name) => [name, entry] as const)));
const WHITESPACE_RE = /\s+/u;

export const SLASH_COMMANDS: SlashCommand[] = registry.flatMap(({ cmd, args, desc, aliases = [] }) => [
  { cmd, args, desc },
  ...aliases,
]);

export function matchSlashCommands(query: string): SlashCommand[] {
  return SLASH_COMMANDS.filter((command) => command.cmd.slice(1).startsWith(query));
}

export function handleSlashCommand(text: string, buffer: Buffer, sendCmd: SendCmd): boolean {
  const [cmd, ...rest] = text.slice(1).split(WHITESPACE_RE);
  const entry = registryByName.get((cmd || "").toLowerCase());
  if (!entry) return false;
  entry.handle({ buffer, rest, args: rest.join(" ").trim(), sendCmd });
  return true;
}

File: apple/Lurker/SlashCommands.swift
import Foundation

// MARK: - ComposerResult

enum ComposerResult: Equatable, Sendable {
  case command(ClientCommand)
  case invalid(String)
}

// MARK: - SlashCommandHelp

struct SlashCommandHelp: Equatable, Identifiable, Sendable {
  let command: String
  let arguments: String
  let description: String

  var id: String {
    command
  }
}

// MARK: - SlashCommands

enum SlashCommands {

  // MARK: Internal

  static let help: [SlashCommandHelp] = [
    .init(command: "/me", arguments: "<action>", description: "Send an action"),
    .init(command: "/join", arguments: "<channel>", description: "Join a channel"),
    .init(command: "/part", arguments: "[reason]", description: "Leave this channel"),
    .init(command: "/query", arguments: "<nick>", description: "Open a conversation"),
    .init(command: "/msg", arguments: "<nick> <message>", description: "Send a direct message"),
    .init(command: "/nick", arguments: "<new nick>", description: "Change your nickname"),
    .init(command: "/whois", arguments: "<nick>", description: "Look up a user"),
    .init(command: "/away", arguments: "[message]", description: "Set away status"),
    .init(command: "/back", arguments: "", description: "Clear away status"),
    .init(command: "/topic", arguments: "[topic]", description: "Set the channel topic"),
    .init(command: "/mode", arguments: "<modes> [params]", description: "Set channel modes"),
    .init(command: "/op", arguments: "<nick>", description: "Give channel op (+o)"),
    .init(command: "/deop", arguments: "<nick>", description: "Remove channel op (-o)"),
    .init(command: "/voice", arguments: "<nick>", description: "Give voice (+v)"),
    .init(command: "/devoice", arguments: "<nick>", description: "Remove voice (-v)"),
    .init(command: "/kick", arguments: "<nick> [reason]", description: "Kick nick from channel"),
    .init(command: "/kickban", arguments: "<nick> [reason]", description: "Kick and ban nick"),
    .init(command: "/ban", arguments: "<mask>", description: "Ban mask (+b)"),
    .init(command: "/unban", arguments: "<mask>", description: "Unban mask (-b)"),
    .init(command: "/banlist", arguments: "", description: "Show channel ban list"),
    .init(command: "/list", arguments: "[filter]", description: "List channels on the network"),
    .init(command: "/archive", arguments: "", description: "Archive this buffer"),
    .init(command: "/unarchive", arguments: "", description: "Unarchive this buffer"),
    .init(
      command: "/delete",
      arguments: "",
      description: "Delete this archived buffer and its history",
    ),
  ]

  /// Commands whose name prefix-matches the first token of a `/`-leading
  /// composer text (web parity: matchSlashCommands — display-only popup, so
  /// the popup stays up while typing arguments as long as the prefix holds).
  static func matching(_ text: String) -> [SlashCommandHelp] {
    guard text.hasPrefix("/") else { return [] }
    let query = text.dropFirst().split(whereSeparator: \.isWhitespace).first.map(String.init) ?? ""
    let lowered = "/" + query.lowercased()
    return help.filter { $0.command.hasPrefix(lowered) }
  }

  static func parse(_ input: String, buffer: Buffer) -> ComposerResult {
    let value = input.trimmingCharacters(in: .whitespacesAndNewlines)
    guard value.hasPrefix("/") else {
      // The status window has no message target — it accepts commands only
      // (web parity: the status composer placeholder says the same).
      if buffer.kind == "status" {
        return .invalid("Commands only here — try /list, /join or /msg.")
      }
      return .command(ClientCommand(type: "send", bufferID: buffer.id, content: value))
    }
    let pieces = value.dropFirst().split(whereSeparator: \.isWhitespace).map(String.init)
    guard let name = pieces.first?.lowercased() else {
      return .invalid("Enter a command.")
    }
    let rest = Array(pieces.dropFirst())
    let content = rest.joined(separator: " ")
    switch name {
    case "me":
      return require(content, usage: "/me <action>") {
        ClientCommand(type: "me", bufferID: buffer.id, content: content)
      }

    case "join":
      return require(content, usage: "/join <channel>") {
        ClientCommand(type: "join", networkID: buffer.networkID, channel: content)
      }

    case "part":
      return .command(ClientCommand(type: "part", bufferID: buffer.id, content: content))

    case "query":
      return require(content, usage: "/query <nick>") {
        ClientCommand(type: "query", networkID: buffer.networkID, target: content)
      }

    case "msg":
      guard let target = rest.first, rest.count > 1 else {
        return .invalid("Usage: /msg <nick> <message>")
      }
      return .command(
        ClientCommand(
          type: "msg",
          networkID: buffer.networkID,
          target: target,
          content: rest.dropFirst().joined(separator: " "),
        )
      )

    case "nick":
      return require(content, usage: "/nick <new nick>") {
        ClientCommand(type: "nick", networkID: buffer.networkID, content: content)
      }

    case "whois":
      return require(content, usage: "/whois <nick>") {
        ClientCommand(type: "whois", networkID: buffer.networkID, target: content)
      }

    case "away":
      return .command(ClientCommand(type: "away", networkID: buffer.networkID, content: content))

    case "back":
      return .command(ClientCommand(type: "back", networkID: buffer.networkID))

    case "topic":
      return .command(ClientCommand(type: "topic", bufferID: buffer.id, content: content))

    case "mode":
      return require(content, usage: "/mode <modes> [params]") {
        ClientCommand(type: "mode", bufferID: buffer.id, content: content)
      }

    case "op",
         "deop",
         "voice",
         "devoice":
      return require(content, usage: "/\(name) <nick>") {
        ClientCommand(type: name, bufferID: buffer.id, target: content)
      }

    case "ban",
         "unban":
      return require(content, usage: "/\(name) <mask>") {
        ClientCommand(type: name, bufferID: buffer.id, target: content)
      }

    case "banlist":
      return .command(ClientCommand(type: "banlist", bufferID: buffer.id))

    case "kick",
         "kickban":
      guard let target = rest.first else {
        return .invalid("Usage: /\(name) <nick> [reason]")
      }
      return .command(
        ClientCommand(
          type: name,
          bufferID: buffer.id,
          target: target,
          content: rest.dropFirst().joined(separator: " "),
        )
      )

    case "list":
      return .command(ClientCommand(type: "list", networkID: buffer.networkID, content: content))

    case "archive":
      guard buffer.kind != "status" else {
        return .invalid("The status window cannot be archived.")
      }
      return .command(ClientCommand(type: "archive_buffer", bufferID: buffer.id))

    case "unarchive":
      guard buffer.kind != "status" else {
        return .invalid("The status window cannot be archived.")
      }
      return .command(ClientCommand(type: "unarchive_buffer", bufferID: buffer.id))

    case "delete":
      guard buffer.kind != "status", buffer.archived else {
        return .invalid("Only archived channels and conversations can be deleted.")
      }
      return .command(ClientCommand(type: "delete_buffer", bufferID: buffer.id))

    default:
      return .invalid("Unknown command /\(name)")
    }
  }

  // MARK: Private

  private static func require(
    _ value: String,
    usage: String,
    command: () -> ClientCommand,
  ) -> ComposerResult {
    value.isEmpty ? .invalid("Usage: \(usage)") : .command(command())
  }

}
